[InterfaceGen] Print property initializers in resilient, fixed-layout types

harlanhaskins opened the pull request:
Augment the ASTPrinter to print the name and text of initializer expressions if
a property has an initializer and the type is `@_fixed_layout` and resides in a resilient module, and serialize the text for partial modules.

With this change, all .swiftinterface files in the project (except for SwiftLang) compile to swiftmodules on macOS.

rdar://43774580
rdar://43812188

harlanhaskins commented:
Still needs a concrete test case exercising all the configurations of resilient/non-resilient @_fixed_layout/non-fixed-layout, inlinable, non-inlinable, default value, non-default value properties.

harlanhaskins commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2011,6 +2026,27 @@ void PrintAST::visitExtensionDecl(ExtensionDecl *decl) {
+    return false;
+  auto parent = dyn_cast<NominalTypeDecl>(vd->getDeclContext());
+  if (!parent) return false;
+  if (parent->isFormallyResilient())
```
@slavapestov This seemed like the right check to do, but is it? I'm trying to see if the type is effectively fixed-layout. Should I be guarding this with a check if the module is resilient?

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2011,6 +2026,27 @@ void PrintAST::visitExtensionDecl(ExtensionDecl *decl) {
+/// There's a very narrow case when we would:
+/// If the decl:
+///   - has an initializer expression
+///   - is @usableFromInline
```
Also if its public

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2011,6 +2026,27 @@ void PrintAST::visitExtensionDecl(ExtensionDecl *decl) {
+///   - is @usableFromInline
+/// If the parent type:
+///   - is @_fixed_layout (or effectively fixed-layout)
+///   - has at least one inlinable initializer in itself or any of its
```
I don't think this second check is necessary

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2011,6 +2026,27 @@ void PrintAST::visitExtensionDecl(ExtensionDecl *decl) {
+///   - has an initializer expression
+///   - is @usableFromInline
+/// If the parent type:
+///   - is @_fixed_layout (or effectively fixed-layout)
```
What does effectively fixed-layout mean? Only types that have an explicit @_fixed_layout attribute need this treatment. If the module is not resilient and the type is implicitly @_fixed_layout, you cannot print the initial value because it might reference private things. However in that case the initializer symbol is public, so you don't need to emit it yourself.

harlanhaskins replied:
If the type is explicitly fixed_layout and the property is non-public but usable from inline, should it still print the value?

slavapestov replied:
Yes, and if the property is public too. @usableFromInline makes things public for ABI purposes but not public in the language itself. So “@usableFromInline but not public” is not a useful category of declaration. It’s always “public” or “public or @usableFromInline”.

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2011,6 +2026,27 @@ void PrintAST::visitExtensionDecl(ExtensionDecl *decl) {
+  if (!parent) return false;
+  if (parent->isFormallyResilient())
+    return false;
+  return evaluateOrDefault(vd->getASTContext().evaluator,
```
I don't understand why this needs to be a request. In fact I don't think you need to check this condition at all.

harlanhaskins replied:
If we are doing the check, it’s nice not to look through the members and extensions for each property of the type. Though, if we don’t need the check all of the request plumbing can go away.

slavapestov commented on test/Constraints/diagnostics.swift:
```diff
@@ -953,7 +953,7 @@ func SR_6272_a() {
   }
 
-  // expected-error@+2 {{binary operator '*' cannot be applied to operands of type 'Int' and 'Float'}} {{35-35=Int(}} {{42-42=)}}
+  // expected-error@+2 {{binary operator '*' cannot be applied to operands of type 'Int' and 'Float'}} {{35-35=Int(}} {{43-43=)}}
```
Why did this diagnostic change?

harlanhaskins replied:
It was previously using an incorrect source location, it was inserting the right paren of the cast before the existing right paren, not after. This was fixed by not removing the outer CoerceExpr when we convert a constructor call to a literal initialization.

slavapestov replied:
Then it should be in a separate PR. Tag @rudkx or @xedin for review.

harlanhaskins committed:
[InterfaceGen] Print property initializers in resilient, fixed-layout types

Augment the ASTPrinter to print the name and text of initializer expressions if
a property has an initializer and the type is @_fixed_layout and resides in a resilient module, and serialize the text for partial modules.

With this change, all .swiftinterface files in the project (except for SwiftLang) compile to swiftmodules on macOS.

rdar://43774580
rdar://43812188

harlanhaskins commented on test/ModuleInterface/fixed-layout-property-initializers.swift:
```diff
@@ -0,0 +1,63 @@
+// RUN: %target-swift-frontend -emit-interface-path %t-resilient.swiftinterface -enable-resilience -emit-module -o /dev/null %s
+// RUN: %FileCheck %s --check-prefix RESILIENT --check-prefix COMMON < %t-resilient.swiftinterface
+
+// FIXME(rdar): %target-swift-frontend -emit-module -o %t/Test.swiftmodule %t.swiftinterface -disable-objc-attr-requires-foundation-module
```
I need to actually file this, hold on

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -1883,6 +1883,9 @@ class PatternBindingEntry {
   DeclContext *InitContext = nullptr;
 
+  /// The text of the initializer expression if deserialized from a module.
+  StringRef InitStringRepresentation;
```
Bitpack nitpick: there are a *lot* of pattern binding entries in the program. Can we cram this in with the initializer expression, and move the flags to the other members?

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -842,26 +842,35 @@ void PrintAST::printAttributes(const Decl *D) {
+    // Don't print @_hasInitialValue if we're in printing an
+    // @usableFromInline property.
+    if (auto vd = dyn_cast<VarDecl>(D)) {
+      if (vd->getAttrs().hasAttribute<UsableFromInlineAttr>())
```
This isn't the right check; it should be the same as how you decide whether to print the initializer expression.

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2011,6 +2023,21 @@ void PrintAST::visitExtensionDecl(ExtensionDecl *decl) {
+  if (!parent) return false;
+  if (!vd->getAttrs().hasAttribute<HasInitialValueAttr>())
+    return false;
+  if (!parent->getAttrs().hasAttribute<FixedLayoutAttr>())
```
You also need to check that it's an instance property.

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -2358,9 +2376,11 @@ bool ValueDecl::isUsableFromInline() const {
-    if (ATD->getProtocol()->getAttrs().hasAttribute<UsableFromInlineAttr>())
+  if (auto *containingProto = dyn_cast<ProtocolDecl>(getDeclContext())) {
+    if (isProtocolRequirement() &&
+        containingProto->getAttrs().hasAttribute<UsableFromInlineAttr>())
```
This looks like one of my changes. I guess it'll get merged away.

harlanhaskins replied:
Ah, yes, this was accidental.

jrose-apple commented on lib/Serialization/Serialization.cpp:
```diff
@@ -2438,9 +2438,16 @@ void Serializer::writePatternBindingInitializer(PatternBindingDecl *binding,
+  StringRef initStr;
+  SmallString<128> scratch;
+  if (binding->hasInitStringRepresentation(bindingIndex)) {
+    initStr = binding->getInitStringRepresentation(bindingIndex, scratch);
```
This should probably not be serialized if we don't need it (the same conditions as printing).

I guess in general we need that configuration option that says whether we're making a partial module or a complete module, which would be good enough.

jrose-apple commented on test/ModuleInterface/fixed-layout-property-initializers.swift:
```diff
@@ -0,0 +1,63 @@
+public struct MyStruct {
+  // CHECK: @_hasInitialValue public var publicVar: [[BOOL:(Swift\.)?Bool]]{{$}}
+  // RESILIENT: public var publicVar: [[BOOL:(Swift\.)?Bool]] = false
+  public var publicVar: Bool = false
```
Nitpick: please include at least one test case where there's a multiple-token initial value, just in case. We also definitely need test cases where multiple variables are initialized by a single initial value, and where there are multiple patterns in a single `var`.

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -1864,76 +1864,104 @@ class ExtensionRange {
+  llvm::PointerIntPair<Pattern *, 3, OptionSet<Flags>> PatternAndFlags;
+
+  /// The location of the equal '=' token.
+  SourceLoc EqualLoc;
```
Do you think we should put this into the union too? It appeals to my "pack everything as small as possible" instincts, but I can also see us doing a better job preserving source locations through deserialization at some point.

harlanhaskins replied:
Do we serialize any SourceLocs currently? I feel like we could get the win here by tying the EqualLoc with the InitExpr, and then pay for it later if we do end up serializing this loc. I don't think it would hurt to throw it in the union now.

jrose-apple replied:
We don't today, on the grounds that the serialized module might be distributed somewhere. Okay then.

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -842,26 +842,35 @@ void PrintAST::printAttributes(const Decl *D) {
+    // Don't print @_hasInitialValue if we're printing an initializer
+    // expression.
+    if (auto vd = dyn_cast<VarDecl>(D)) {
+      if (vd->shouldSerializeInitializerText())
```
I wonder if this indicates that the name shouldn't mention "serialization". What do you think of "isInitExposedToClients" or something? That would also make sense as the condition being checked in Sema when deciding whether to enforce inlinable-like rules.

harlanhaskins replied:
Yeah, I tried to find a better name that doesn't reference serialization or printing, but I didn't want the name to be something like `hasInitInResilientFixedLayoutType`. I like `isInitExposedToClients`.

jrose-apple commented on test/ModuleInterface/fixed-layout-property-initializers.swift:
```diff
@@ -0,0 +1,67 @@
+// RUN: %FileCheck %s --check-prefix RESILIENT --check-prefix COMMON < %t-resilient.swiftinterface
+
+// FIXME(rdar44993525): %target-swift-frontend -emit-module -o %t/Test.swiftmodule %t.swiftinterface -disable-objc-attr-requires-foundation-module
+// FIXME(rdar44993525): %target-swift-frontend -emit-module -o /dev/null -merge-modules %t/Test.swiftmodule -module-name Test -emit-interface-path - | %FileCheck %s --check-prefix CHECK --check-prefix COMMON
```
For my own curiosity, what goes wrong here?

harlanhaskins replied:
Since we don't print initializers in a non-resilient module, compiling the emitted interface still fails with:

```
error: return from initializer without initializing all stored properties
```

slavapestov replied:
The presence of the @hasInitialValue attribute should be the same as an initializer in a non-resilient module, since the initializer expression has a public entry point in the dylib

jrose-apple commented on test/ModuleInterface/fixed-layout-property-initializers.swift:
```diff
@@ -0,0 +1,67 @@
+// RUN: %empty-directory(%t)
+
+// RUN: %target-swift-frontend -emit-interface-path %t.swiftinterface -emit-module -o /dev/null %s
```
I merged my other change, so you can switch these to `-typecheck` if you want.

jrose-apple commented on test/ModuleInterface/fixed-layout-property-initializers.swift:
```diff
@@ -0,0 +1,67 @@
+
+  // COMMON: @inlinable internal init() {}
+  @inlinable init() {}
+}
```
Missing a test case for `static` and for top-level properties.

jrose-apple reviewed (commented):
Looks pretty good. All minor comments left.

harlanhaskins commented:
@swift-ci please smoke test

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -899,7 +908,10 @@ void PrintAST::printPattern(const Pattern *pattern) {
-          !isPublicOrUsableFromInline(decl))
+          !isPublicOrUsableFromInline(decl) &&
+          // FIXME: We need to figure out a way to generate an entry point
+          //        for the initializer expression without revealing the name.
```
I don't think that's possible, since the mangled name of the initializer contains the property name

jrose-apple replied:
We don't actually care what the name is, though, since it's not a public entry point. At least, we don't for resilient modules.

slavapestov replied:
If we generate a name here it has to be consistent between invocations though, so that you can do a non-WMO build where multiple source files import the same textual interface (and the cached binary form got rebuild between invocations)

slavapestov replied:
I would again advocate for the simpler approach, instead of figuring out the exact strategy for when we can elide the name, just be conservative and only elide it if there's no initializer

jrose-apple replied:
We're talking about a swiftinterface here; it's always one file.

jrose-apple replied:
And we can't just print the name because the name might conflict with extensions.

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test

harlanhaskins commented:
@swift-ci please smoke test
